scene/3d/ray_cast.cpp

```diff
@@ -119,13 +132,19 @@ void RayCast::_notification(int p_what) {
 				set_fixed_process(false);
 			}
 
+			if (debug_shape)
+				_clear_debug_shape();
+
 		} break;
 		case NOTIFICATION_FIXED_PROCESS: {
 
 			if (!enabled)
 				break;
 
+			bool prev_collision_state = collided;
 			_update_raycast_state();
+			if (prev_collision_state != collided && get_tree()->is_debugging_collisions_hint())
+				_update_debug_shape();
```

Comment:
I'm not sure about this, you are adding a very expensive call again in the FIXED_PROCESS. In case the collision changes, all that is needed to do is change the color, not re-create the whole mesh